main/src/main/scala/sbt/internal/XMainConfiguration.scala

```diff
@@ -10,10 +10,43 @@ package sbt.internal
 import java.io.File
 import java.lang.reflect.InvocationTargetException
 import java.net.{ URL, URLClassLoader }
+import java.util.concurrent.{ ExecutorService, Executors }
 import java.util.regex.Pattern
 
+import sbt.plugins.{ CorePlugin, IvyPlugin, JvmPlugin }
+import sbt.util.LogExchange
 import xsbti._
 
+private[internal] object ClassLoaderWarmup {
+  def warmup(): Unit = {
+    if (Runtime.getRuntime.availableProcessors > 1) {
+      val executorService: ExecutorService =
+        Executors.newFixedThreadPool(Runtime.getRuntime.availableProcessors - 1)
+      def submit[R](f: => R): Unit = {
+        executorService.submit(new Runnable {
+          override def run(): Unit = { f; () }
+        })
+        ()
+      }
+
+      submit(LogExchange.context)
+      submit(Class.forName("sbt.internal.parser.SbtParserInit").getConstructor().newInstance())
+      submit(CorePlugin.projectSettings)
+      submit(IvyPlugin.projectSettings)
+      submit(JvmPlugin.projectSettings)
+      submit(() => {
+        try {
+          val clazz = Class.forName("scala.reflect.runtime.package$")
+          clazz.getMethod("universe").invoke(clazz.getField("MODULE$").get(null))
+        } catch {
+          case _: Exception =>
+        }
+        executorService.shutdown()
```

Comment:
Btw, this needn't be here, within the job submission.  I'm assuming the logic was all the other jobs will be done before the runtime universe is down initialising.

Instead, it could be just after the last job submission (i.e. 1 line down here), because `shutdown` works by triggering the service to not accept any new submissions, but completing all the ones in flight.  (`awaitTermination` does the same but then also blocks, waiting for the shutdown to complete.)

Not worth PRing, IMO, as nothing of substance will be gained doing the shutdown earlier in the main thread instead of later in this warm up runtime universe thread.  Just FYI.